py/torch_tensorrt/runtime/_cudagraphs.py

```diff
@@ -37,19 +64,52 @@ class _CudagraphsContextManager(object):
     Used to enable cudagraphs as a context manager
     """
 
-    def __init__(self) -> None:
+    def __init__(self, compiled_module: torch.nn.Module) -> None:
         global _PY_RT_CUDAGRAPHS
         self.old_mode = _PY_RT_CUDAGRAPHS
+        self.compiled_module = compiled_module
+
+    def __enter__(self) -> torch.nn.Module:
+        global _PY_RT_CUDAGRAPHS
 
-    def __enter__(self) -> "_CudagraphsContextManager":
-        # Enable cudagraphs
-        set_cudagraphs_mode(True)
-        return self
+        num_torch_module = 0
+        num_trt_module = 0
+        for name, _ in self.compiled_module.named_children():
+            if "_run_on_acc" in name:
+                num_trt_module += 1
+            elif "_run_on_gpu" in name:
+                num_torch_module += 1
+
+        if num_torch_module > 0:
+            # Set whole cudagraphs mode and returns wrapped module
+            _PY_RT_CUDAGRAPHS = CudaGraphsMode.WHOLE_GRAPH_CUDAGRAPHS
+            # Set new mode for C++
+            if torch_tensorrt.ENABLED_FEATURES.torch_tensorrt_runtime:
+                torch.ops.tensorrt.set_cudagraphs_mode(_PY_RT_CUDAGRAPHS)
+
+            logger.debug(
+                f"{num_torch_module} torch modules are in subgraphs. Using wrapper module for cuda graphs"
+            )
+            return CudaGraphsTorchTensorRTModule(self.compiled_module)
+        else:
+            if num_trt_module > 0:
+                logger.debug(
+                    "There is no graph breaks. Using original module for cuda graphs"
+                )
+            else:
```

Comment:
If `torch_compile` backend is used, graph information is not available when context manager is generated.
Context manager will return original module in this case. Documented that `dynamo` backend is required for cuda graphs+ graphs breaks.
https://github.com/pytorch/TensorRT/pull/3270/files#diff-7f89360445589890395e2fa77a673994a449f464d82482f51f1df3c448700a04R96

Reply:
Also, so is this saying that if the user is using torch.compile that we will use subgraph_cudagraphs? I think this is probably fine not sure if it needs to be a warning but we should print in the debug logs which one we choose.